File: src/main/GameLogic.cpp

#define SIZE 16

enum class Direction
{
  UP,
  DOWN,
  LEFT,
  RIGHT
};
enum class Side
{
  FRONT,
  LEFT,
  TOP,
  RIGHT
};

struct CellPos
{
  Side side;
  int x, y;

  bool operator==(const CellPos &other) const
  {
    return side == other.side && x == other.x && y == other.y;
  }

  CellPos moveClone(Direction direction)
  {
    CellPos newPos = {side, x, y};
    newPos.moveSelf(direction);
    return newPos;
  }

  void moveSelf(Direction direction)
  {
    CellPos newPos = {side, x, y};
    switch (direction)
    {
    case Direction::UP:
      if (y >= SIZE - 1)
      {
        if (side == Side::FRONT)
        {
          side = Side::TOP;
          y = 0;
        }
        else if (side == Side::LEFT)
        {
          side = Side::TOP;
          y = SIZE - x - 1;
          x = 0;
        }
        else if (side == Side::RIGHT)
        {
          side = Side::TOP;
          y = x;
          x = SIZE - 1;
        }
        side = Side::TOP;
      }
      else
      {
        y++;
      }
      break;
    case Direction::DOWN:
      if (y <= 0)
      {
        if (side == Side::TOP)
        {
          side = Side::FRONT;
          y = SIZE - 1;
        }
      }
      else
      {
        y--;
      }
      break;
    case Direction::LEFT:
      if (x <= 0)
      {
        if (side == Side::FRONT)
        {
          side = Side::LEFT;
          x = SIZE - 1;
        }
        else if (side == Side::TOP)
        {
          side = Side::LEFT;
          x = SIZE - y - 1;
          y = SIZE - 1;
        }
        else if (side == Side::RIGHT)
        {
          side = Side::FRONT;
          x = SIZE - 1;
        }
      }
      else
      {
        x--;
      }
      break;
    case Direction::RIGHT:
      if (x >= SIZE - 1)
      {
        if (side == Side::FRONT)
        {
          side = Side::RIGHT;
          x = 0;
        }
        else if (side == Side::TOP)
        {
          side = Side::RIGHT;
          x = y;
          y = SIZE - 1;
        }
        else if (side == Side::LEFT)
        {
          side = Side::FRONT;
          x = 0;
        }
      }
      else
      {
        x++;
      }
      break;
    }
  }
};

struct Player
{
  CellPos pos;
  Direction direction;
};

struct CellPlayerData
{
  bool atPosition, activeTrail, ownedBy;
};

class GameLogic
{
public:
  GameLogic()
  {
    initArrays();
    updateMatrix();
  }
  ~GameLogic()
  {
    delete[] data;
    delete[] players;
  }

  void tick()
  {
    movePlayers();
    updateMatrix();
  }

  CellPlayerData *getCellPlayerData(CellPos pos, int p)
  {
    return &data[pos.y * SIZE + pos.x + p * SIZE * SIZE];
  }

  Player *getPlayer(int p)
  {
    return &players[p];
  }

private:
  int numPlayers = 2;
  CellPlayerData *data;
  Player *players;

  void initArrays()
  {
    int cellCount = SIZE * SIZE * numPlayers;
    data = new CellPlayerData[cellCount];
    for (int i = 0; i < cellCount; i++)
      data[i] = {false, false, false};
    players = new Player[2];
    players[0] = {{Side::FRONT, 0, 0}, Direction::RIGHT};
    players[1] = {{Side::FRONT, 1, 1}, Direction::LEFT};
  }

  void movePlayers()
  {
    for (int i = 0; i < numPlayers; i++)
    {
      Player *player = &players[i];
      player->pos.moveSelf(player->direction);
    }
  }

  void updateMatrix()
  {
    for (int y = 0; y < SIZE; y++)
    {
      for (int x = 0; x < SIZE; x++)
      {
        for (int p = 0; p < numPlayers; p++)
        {
          CellPos pos = {Side::FRONT, x, y};
          updateCell(pos, p);
        }
      }
    }
  }

  void updateCell(CellPos pos, int p)
  {
    Player *player = getPlayer(p);
    CellPlayerData *cell = getCellPlayerData(pos, p);
    cell->atPosition = pos == player->pos;
    if (cell->atPosition && !cell->activeTrail)
    {
      cell->activeTrail = true;
    }
  }
};

File: py/game_logic.py
from enums import Direction
from side import Side
from cell_pos import CellPos

class Player:
    def __init__(self, pos, direction):
        self.pos = pos
        self.direction = direction

class PlayerCellData:
    def __init__(self):
        self.at_position = False
        self.active_trail = False

class GameLogic:
    def __init__(self, size):
        self.size = size
        self.cells = {}
           
        front = Side("front", self.size)
        top = Side("top", self.size)
        left = Side("left", self.size)
        right = Side("right", self.size)
        self.players = [
            Player(CellPos(front, 0, 0), Direction.RIGHT),
            # Player(CellPos(front, size - 1, size - 1), Direction.LEFT)
        ]

        front.attach_top(top, Direction.DOWN)
        front.attach_left(left, Direction.RIGHT)
        front.attach_right(right, Direction.LEFT)

        top.attach_down(front, Direction.UP)
        top.attach_left(left, Direction.UP)
        top.attach_right(right, Direction.UP)

        left.attach_right(front, Direction.LEFT)
        left.attach_top(top, Direction.LEFT)

        right.attach_left(front, Direction.RIGHT)
        right.attach_top(top, Direction.RIGHT)

        self.sides = [front, left, right, top]

        for side in self.sides:
            for x in range(self.size):
                for y in range(self.size):
                    pos = CellPos(side, x, y)
                    self.cells[pos] = [PlayerCellData() for _ in range(len(self.players))]

    def tick(self):
        self.move_players()
        self.update_matrix()

    def get_cell_data(self, pos):
        return self.cells[pos]

    def get_player(self, p):
        return self.players[p]

    def move_players(self):
        for player in self.players:
            player.pos.move(player.direction)

    def update_matrix(self):
       for pos in self.cells: 
            self.update_cell(pos, self.cells[pos])

    def update_cell(self, pos, cell):
        for i in range(len(self.players)): 
            player = self.players[i]
            cell[i].at_position = pos == player.pos
            if cell[i].at_position and not cell[i].active_trail:
                cell[i].active_trail = True
